### PR TITLE
Init before login if not initialized

### LINKED ISSUE #243: A more intuitive error exists when user attempts to login before initializing the CLI
## Summary
When the .conjurrc does not exist (created during `conjur init`) and the user attempts to login (`conjur login`), they get the error `Error: No such file or directory: '/Users/<>/.conjurrc'`. This is not a great user experience because it does not tell the user what has happened or why it has happened

## Steps to Reproduce
1. Remove `.conjurrc` if exists on your machine
2. Run `conjur login -i admin -p <somepass>`
3. See error `Error: No such file or directory: '/Users/<>/.conjurrc'`

## Expected Results
One of following:
1. Prompt them to initialize the client (run `init` and then run the remaining command (`conjur login`)
2. Return informative error `Error: You have not initalized the client. Run "conjur init" and try again.`
3. Any other options you see fit

## Reproducible
   * [x] Always 
   * [ ] Sometimes
   * [ ] Non-Reproducible

## Environment setup
Dev environment on local macOS machine. See [here](https://github.com/cyberark/conjur-api-python3/blob/master/CONTRIBUTING.md#development)

## Additional Information
The flow should be smart enough to cover when a user provides flags and when they don't. In other words it would be great to have some sort of input validation.

For example, the following immediately prompts the error
`./pkg_bin/conjur login -i admin -p <somepass>`

BUT the following does not and requests user interaction
```
./pkg_bin/conjur login
Enter your login name to log into Conjur:
```

Some input validation already exists but this might not be the proper place. Consider taking a look at this for inspiration and see there is a better place for all input validations: https://github.com/cyberark/conjur-api-python3/blob/master/conjur/cli.py#L727

## DOD
- [ ] Implementation has been done
- [ ] Integration / Unit tests have been written
- [ ] Live demo addition
- [ ] All necessary logs have been written and approved by TW / PO

### PR DESCRIPTION
**Pull request by eranha**
### What does this PR do?
- Complete the work on issue #243
- Nest function call under an if condition 
- Make tests pass

### What ticket does this PR close?
Resolves #243 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation

### CODE REVIEW

File: conjur/cli.py [hunk 1]
```diff
@@ -706,33 +707,38 @@ def run_action(resource, args):
...
-            # pylint: disable=line-too-long
-            if not os.path.exists(DEFAULT_CONFIG_FILE) or os.path.getsize(DEFAULT_CONFIG_FILE) == 0:
-                sys.stdout.write("Error: The Conjur CLI has not been initialized\n")
-                Cli.handle_init_logic()
-
-            # If the user runs a command without logging into the CLI,
-            # we request they do so before executing their request
+        # If the user runs a command without logging into the CLI,
+        # we request they do so before executing their request
+
```
Comment (sigalsax) [inline conjur/cli.py:737]: remove empty line

File: conjur/cli.py [hunk 2]
```diff
@@ -706,33 +707,38 @@ def run_action(resource, args):
...
+        if not is_testing_env:
             loaded_conjurrc = ConjurrcData.load_from_file()
             if not credential_provider.is_exists(loaded_conjurrc.conjur_url):
-                sys.stdout.write("Error: You have not logged in\n")
+                sys.stdout.write("To start using the CLI, log in to Conjur.\n")
```
Comment (sigalsax) [inline conjur/cli.py:741]: remove the period at the end of the sentence. We don't want a hard stop

File: conjur/controller/init_controller.py [hunk 3]
```diff
@@ -53,8 +53,8 @@ def load(self):
         self.get_account_info(self.conjurrc_data)
         self.write_conjurrc()
 
-        sys.stdout.write("Configuration was initialized successfully.\n")
-        sys.stdout.write("To begin using the Conjur CLI, log in to the Conjur server by "
+        sys.stdout.write("Successfully initialized the Conjur CLI.\n")
```
Comment (sigalsax) [inline conjur/controller/init_controller.py:56]: remove the period at the end of the sentence. We don't want a hard stop

File: conjur/controller/init_controller.py [hunk 4]
```diff
@@ -89,10 +89,10 @@ def get_server_certificate(self):
         fingerprint, fetched_certificate = self.init_logic.get_certificate(url.hostname, url.port)
 
         sys.stdout.write(f"\nThe Conjur server's certificate SHA-1 fingerprint is:\n{fingerprint}\n")
-        sys.stdout.write("\nTo verify this certificate, it is recommended to run the following "
+        sys.stdout.write("\nTo verify this certificate, we recommend running the following "
                          "command on the Conjur server:\n"
                          "openssl x509 -fingerprint -noout -in ~conjur/etc/ssl/conjur.pem\n\n")
-        trust_certificate = input("Trust this certificate? (Default=no): ").strip()
+        trust_certificate = input("Trust this certificate? yes/no (Default: no): ").strip()
```
Comment (sigalsax) [inline conjur/controller/init_controller.py:95]: Please update other parts of the codebase that has this (Default=) to ensure uniformity. For example for all file overwrites `(Overwrite? (Default=yes)`, `Trust this certificate? (Default=no)` etc

File: conjur/util/util_functions.py [hunk 5]
```diff
@@ -19,3 +20,11 @@ def determine_status_code_specific_error_messages(server_error):
         return "Failed to log in to Conjur. Unable to authenticate with Conjur. " \
               f"Reason: {server_error}. Check your credentials and try again.\n"
     return f"Failed to execute command. Reason: {server_error}\n"
+
```
Comment (sigalsax) [inline conjur/util/util_functions.py:23]: remove empty line

File: test/test_integration_variable.py [hunk 6]
```diff
@@ -269,7 +269,7 @@ def test_variable_get_without_user_logged_in_prompts_login_and_performs_get(self
             output = self.invoke_cli(self.cli_auth_params,
                                      ['variable', 'set', '-i', 'one/password', '-v', 'somevalue'])
 
-            self.assertIn("Error: You have not logged in", output)
+            self.assertIn("To start using the CLI, log in to Conjur.", output)
```
Comment (sigalsax) [inline test/test_integration_variable.py:272]: remove period at the end

File: test/test_unit_cli.py [hunk 7]
```diff
@@ -289,14 +289,15 @@ def test_run_action_runs_init_logic(self, mock_setup_logging, mock_handle_init):
...
+    @patch.object(FileCredentialsProvider, 'is_exists', return_value=True)
     @patch.object(Cli, 'handle_init_logic')
     @patch.object(Cli, 'handle_variable_logic')
     @patch('os.path.exists', side_effect=[False, True])
     @patch('os.getenv', return_value=None)
     @patch('os.path.getsize', side_effect=[1, 1])
     @patch('conjur.data_object.conjurrc_data.ConjurrcData.load_from_file', return_value=MockConjurrc)
     @patch('keyring.get_keyring')
-    def test_run_action_runs_init_if_conjurrc_not_found(self, mock_keyring, mock_conjurrc, mock_size, mock_getenv, mock_path_exists, mock_variable_init, mock_handle_init):
+    def test_run_action_runs_init_if_conjurrc_not_found(self, mock_keyring, mock_conjurrc, mock_size, mock_getenv,    mock_path_exists, mock_variable_init, mock_handle_init, mock_exists):
```
Comment (sigalsax) [inline test/test_unit_cli.py:300]: remove extra space here before `mock_path_exists`

File: test/test_unit_ssl_service.py [hunk 8]
```diff
@@ -3,61 +3,60 @@
 
 from conjur.api.ssl_client import SSLClient
 
-GITHUB_FINGERPRINT = "54:C5:A2:27:06:92:CF:D1:1C:24:A3:92:96:41:49:06:21:B0:45:30"
+GITHUB_FINGERPRINT = "84:63:B3:A9:29:12:CC:FD:1D:31:47:05:98:9B:EC:13:99:37:D0:D7"
```
Comment (sigalsax) [inline test/test_unit_ssl_service.py:6]: As part of this failing fingerprint test we could fetch the certificate as part of the test (you could fetch the cert as part of the test `openssl s_client -connect github.com:443 < /dev/null 2>/dev/null | openssl x509 -fingerprint -noout -in /dev/stdin`) and compare that to whatever the CLI fetches

File: conjur/cli.py [hunk 9]
```diff
@@ -706,33 +707,38 @@ def run_action(resource, args):
 
         credential_provider, _ = CredentialStoreFactory.create_credential_store()
         # pylint: disable=no-else-return,line-too-long
+        if resource == 'logout':
```
Comment (sigalsax) [inline conjur/cli.py:710]: why was this placed here instead of below ? Does this make a difference?

File: conjur/cli.py (hunk 2 above)
Comment (eranha) [reply conjur/cli.py:741]: I think that is what Shuli gave Abraham

File: test/test_integration_variable.py (hunk 6 above)
Comment (eranha) [reply test/test_integration_variable.py:272]: That's why I prefer const

File: conjur/cli.py (hunk 2 above)
Comment (sigalsax) [reply conjur/cli.py:741]: that's fine. Shuli told me to remove periods

File: test/test_unit_ssl_service.py (hunk 8 above)
Comment (eranha) [reply test/test_unit_ssl_service.py:6]: You're assuming that openssl is present

File: conjur/cli.py (hunk 9 above)
Comment (eranha) [reply conjur/cli.py:710]: where below?

File: test/test_unit_ssl_service.py (hunk 8 above)
Comment (sigalsax) [reply test/test_unit_ssl_service.py:6]: in the test container there should be if not, add the dependency or find another way to fetch cert (:

File: conjur/cli.py (hunk 9 above)
Comment (sigalsax) [reply conjur/cli.py:710]: where it was

File: conjur/cli.py [hunk 10]
```diff
@@ -706,33 +707,37 @@ def run_action(resource, args):
...
-        elif resource == 'login':
+
+        # Needed for unit tests so that they do not require configuring
+        is_testing_env = os.getenv('TEST_ENV') in ('true', 'True')
+
+        # If the user runs a command without configuring the CLI,
+        # we request they do so before executing their request
+        # pylint: disable=line-too-long
+        if not is_testing_env and file_is_missing_or_empty(DEFAULT_CONFIG_FILE):
+            sys.stdout.write("The Conjur CLI needs to be initialized before you can use it.\n")
```
Comment (sigalsax) [inline conjur/cli.py:728]: remove period

File: conjur/cli.py [hunk 11]
```diff
@@ -706,33 +707,37 @@ def run_action(resource, args):
...
-
-            # If the user runs a command without logging into the CLI,
-            # we request they do so before executing their request
+        # If the user runs a command without logging into the CLI,
+        # we request they do so before executing their request
+        if not is_testing_env:
             loaded_conjurrc = ConjurrcData.load_from_file()
             if not credential_provider.is_exists(loaded_conjurrc.conjur_url):
-                sys.stdout.write("Error: You have not logged in\n")
+                sys.stdout.write("%s\n" % LOGIN_IS_REQUIRED)
```
Comment (sigalsax) [inline conjur/cli.py:740]: %s or fstring?

File: conjur/controller/init_controller.py [hunk 12]
```diff
@@ -53,8 +53,8 @@ def load(self):
         self.get_account_info(self.conjurrc_data)
         self.write_conjurrc()
 
-        sys.stdout.write("Configuration was initialized successfully.\n")
-        sys.stdout.write("To begin using the Conjur CLI, log in to the Conjur server by "
+        sys.stdout.write("Successfully initialized the Conjur CLI\n")
+        sys.stdout.write("To start using the Conjur CLI, log in to the Conjur server by "
```
Comment (sigalsax) [inline conjur/controller/init_controller.py:57]: can't you use LOGIN_IS_REQUIRED here?

File: conjur/constants.py [hunk 13]
```diff
@@ -36,3 +36,5 @@
 
 # For testing purposes
 TEST_HOSTNAME = "https://conjur-https"
+
```
Comment (sigalsax) [inline conjur/constants.py:39]:
```suggestion
# For user interaction
```

File: conjur/constants.py [hunk 14]
```diff
@@ -36,3 +36,5 @@
...
+LOGIN_IS_REQUIRED = "To start using the CLI, log in to Conjur"
```
Comment (sigalsax) [inline conjur/constants.py:40]: When I run `conjur list` before init and login, I am prompted to initialize. At the end I get 
```
Successfully initialized the Conjur CLI
To start using the Conjur CLI, log in to the Conjur server by running `conjur login`
To start using the CLI, log in to Conjur
Enter your username:
```

Do you think this double "To start using" looks ok?

File: conjur/controller/init_controller.py [hunk 15]
```diff
@@ -159,6 +159,6 @@ def ensure_overwrite_file(cls, config_file):
         Method to handle user overwriting logic
         """
         force_overwrite = input(f"File {config_file} exists. "
-                                f"Overwrite? (Default=yes): ").strip()
+                                f"Overwrite? yes/no (Default: yes): ").strip()
```
Comment (sigalsax) [inline conjur/controller/init_controller.py:162]: You missed `Trust this certificate? yes/no (Default=no)`. This should be Default: no

File: conjur/cli.py (hunk 11 above)
Comment (sigalsax) [reply conjur/cli.py:740]: do you think we should do this for every user interaction string?

File: conjur/constants.py (hunk 13 above)
Comment (sigalsax) [reply conjur/constants.py:39]: as in add the comment above the constant you added

File: conjur/cli.py (hunk 11 above)
Comment (sigalsax) [reply conjur/cli.py:740]: sys.stdout.write(f"{LOGIN_IS_REQUIRED}\n")

pylint will probably fail so add `# pylint: disable=logging-fstring-interpolation` above

File: conjur/constants.py (hunk 14 above)
Comment (sigalsax) [reply conjur/constants.py:40]: `Successfully initialized the Conjur CLI.` remove period

File: conjur/cli.py [hunk 16]
```diff
@@ -737,7 +739,8 @@ def run_action(resource, args):
         if not is_testing_env:
             loaded_conjurrc = ConjurrcData.load_from_file()
             if not credential_provider.is_exists(loaded_conjurrc.conjur_url):
-                sys.stdout.write("%s\n" % LOGIN_IS_REQUIRED)
+                # pylint: disable=logging-fstring-interpolation
+                sys.stdout.write(f"{LOGIN_IS_REQUIRED}\n")
```
Comment (sigalsax) [inline conjur/cli.py:743]: if you decide against my [above comment](https://github.com/cyberark/conjur-api-python3/pull/244/commits/a7b0be00a59dac79e94b75d298a223408596489a#r604609192) then let's also extract the second log message from the init_controller.py ("Successfully initialized the Conjur CLI") and lets join that with the above line so it becomes

sys.stdout.write(f"Successfully initialized the Conjur CLI. To start using the CLI, log in to Conjur\n")

File: conjur/cli.py [hunk 17]
```diff
@@ -716,6 +716,8 @@ def run_action(resource, args):
             Cli.handle_init_logic(args.url, args.name, args.certificate, args.force, args.ssl_verify)
             # A successful exit is required to prevent the initialization of
             # the Client because the init command does not require the Client
+            sys.stdout.write("To start using the Conjur CLI, log in to the Conjur server by "
```
[the remark this answers is not in the document]
Comment (eranha) [reply conjur/cli.py:719]: I agree but I did that because of the list issue

**Review by sigalsax — APPROVED** [review]
LGTM!

File: conjur/cli.py [hunk 18]
```diff
@@ -737,7 +739,8 @@ def run_action(resource, args):
         if not is_testing_env:
             loaded_conjurrc = ConjurrcData.load_from_file()
             if not credential_provider.is_exists(loaded_conjurrc.conjur_url):
-                sys.stdout.write("%s\n" % LOGIN_IS_REQUIRED)
+                # pylint: disable=logging-fstring-interpolation
+                sys.stdout.write(f"{LOGIN_IS_REQUIRED}\n")
                 Cli.handle_login_logic(credential_provider, ssl_verify=args.ssl_verify)
```
Comment (sigalsax) [inline conjur/cli.py:747]: There are two places in the KeyringAdapter that do not have tests for. Do you see value in adding tests for this as well?

Screen Shot 2021-03-31 at 12 57 00 PM

**Pull request merged**